File: lib/request.js
var mongoose = require('mongoose');

//https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex
function escapeRegExp(str) {
    return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
}

function listartists(filter, callback, skip, limit){
    var Track = mongoose.model('track');
    var stages = [
        {$match: filter},
        {$group: {_id: {artist: '$artist'}}},
        {$project: {_id: 0, artist: '$_id.artist'}},
        {$sort: {artist: 1}}
    ];
    if (skip) stages.push({$skip: skip});
    if (limit) stages.push({$limit: limit});
    Track.aggregate(
        stages,
        callback
    );
}

function listgenres(filter, callback, skip, limit){
    var Track = mongoose.model('track');
    var stages = [
        {$match: filter},
        {$group: {_id: {genre: '$genre'}}},
        {$project: {_id: 0, genre: '$_id.genre'}},
        {$sort: {genre: 1}}
    ];
    if (skip) stages.push({$skip: skip});
    if (limit) stages.push({$limit: limit});
    Track.aggregate(
        stages,
        callback
    );
}

function listalbums(filter, callback, skip, limit){
    var Track = mongoose.model('track');
    var stages = [
        {$match: filter},
        {$group: {_id: {artist: '$artist', album: '$album'}, year: {$first: '$year'}, artist: {$first: '$artist'}}},
        {$sort: {artist: 1, year: -1, album: 1}},
        {$project: {_id: 0, artist: 1, year: 1, album: '$_id.album'}}
    ];
    if (skip) stages.push({$skip: skip});
    if (limit) stages.push({$limit: limit});
    Track.aggregate(
        stages,
        callback
    );
}

/**
 * JSON list of albums by artists.
 * e.g.
 *   [{
 *     artist: "artist1 name",
 *     albums: [
 *       {album: "album1 name", albumart: "http://mydomain.com/album1/art"},
 *       {album: "album2 name"},
 *     ]
 *   }, ...]
 */
function listalbumsbyartists(filter, callback, skip, limit){
    var albumsByArtists = [],
        lastArtist = null,
        Track = mongoose.model('track');
        
    var stages = [
        {$match: filter},
        {$group: {_id: {artist: '$artist', album: '$album'}, year: {$first: '$year'}, artist: {$first: '$artist'}, first_path: {$first: '$path'}}},
        {$sort: {artist: 1, year: -1, album: 1}},
        {$project: {_id: 0, artist: 1, year: 1, album: '$_id.album'}}
    ];
    if (skip) stages.push({$skip: skip});
    if (limit) stages.push({$limit: limit});

    // Fetch tracks info
    Track.aggregate(
        stages,
        function(err, docs) {
            if (err){
                callback(err, null);
            } else {
                for (var i=0; i<docs.length; i++){
                    var album = {name: docs[i].album, year: docs[i].year},
                        artist = docs[i].artist;
                    album.albumart = 'albumart?album=' + encodeURIComponent(docs[i].album) + '&artist=' + encodeURIComponent(docs[i].artist);
                    if (artist !== null){
                        if (lastArtist !== artist){
                            // add artist with first album
                            albumsByArtists.push({artist: artist, albums: [album]});
                            lastArtist = artist;
                        } else {
                            // artist already exists, so only add album to it
                            albumsByArtists[albumsByArtists.length-1].albums.push(album);
                        }
                    }
                }
                callback(null, albumsByArtists);
            }
        }
    );
}

function listtracks(filter, callback, skip, limit) {
    var Track = mongoose.model('track');
    var query = Track.find(filter).sort({ artist: 1, year: -1, album: 1, trackno: 1 });
    if (skip) query.skip(skip);
    if (limit) query.limit(limit);
    query.exec(callback);
}

function fileinfo(ids, callback) {
    listtracks({ _id : {$in: ids}}, callback);
}

function searchartists(term, callback, skip, limit) {
    term = escapeRegExp(term);
    var filter = {artist : {$regex: new RegExp(term, 'i')}};
    listartists(filter, callback, skip, limit);
}

function searchalbums(term, callback, skip, limit) {
    term = escapeRegExp(term);
    var filter = {album : {$regex: new RegExp(term, 'i')}};
    listalbums(filter, callback, skip, limit);
}

function search(term, filters, callback, skip, limit) {
    if (filters) {
        term = escapeRegExp(term);
        var filter = {
            $or: []
        };
        if (filters) {
            if (filters.artists) {
                filter.$or.push({artist : {$regex: new RegExp(term, 'i')}});
            }
            if (filters.albums) {
                filter.$or.push({album : {$regex: new RegExp(term, 'i')}});
            }
            if (filters.tracks) {
                filter.$or.push({name : {$regex: new RegExp(term, 'i')}});
            }
        }
        listtracks(filter, callback, skip, limit);
    } else {
        callback("No filter specified", []);
    }
}

function albumart(artist, album, callback){
    var Albumart = mongoose.model('albumart');
    var query = Albumart.findOne({album: album, artist: artist});
    query.exec(callback);
}

module.exports = {
    listartists: listartists,
    listgenres: listgenres,
    listalbums: listalbums,
    listalbumsbyartists: listalbumsbyartists,
    listtracks: listtracks,
    fileinfo: fileinfo,
    searchartists: searchartists,
    searchalbums: searchalbums,
    search: search,
    albumart: albumart
};

File: lib/request.py
from .model import session_scope, Track, Album, Artist, Genre
from sqlalchemy import or_
from sqlalchemy.orm import joinedload, joinedload_all, contains_eager, aliased

def limitoffset(query, skip, limit):
    if skip is not None:
        query = query.offset(skip)
    if limit is not None:
        query = query.limit(limit)
    return query

def getartist(artist_id):
    with session_scope() as session:
        obj = session.query(Artist).get(artist_id)
        session.expunge_all()
        return obj

def getalbum(album_id=None, ffilter=None):
    with session_scope() as session:
        if album_id is not None:
            obj = session.query(Album).get(album_id)
            session.expunge_all()
            return obj
        elif ffilter is not None:
            query = session.query(Album).join(Album.artist).options(contains_eager(Album.artist))
            query = ffilter(query)
            obj = query.one()
            session.expunge_all()
            return obj
    return None

def gettrack(track_id=None, ffilter=None):
    with session_scope() as session:
        if track_id is not None:
            obj = session.query(Track).get(track_id)
            session.expunge_all()
            return obj
        elif ffilter is not None:
            query = session.query(Track).join(Track.album).join(Album.artist).options(joinedload(Track.genre), contains_eager(Track.album, Album.artist))
            query = ffilter(query)
            obj = query.one()
            session.expunge_all()
            return obj
    return None

def gettrackfull(track_id):
    with session_scope() as session:
        obj = session.query(Track).join(Track.album).join(Album.artist).options(joinedload(Track.genre), contains_eager(Track.album, Album.artist)).filter(Track.id == track_id).one()
        session.expunge_all()
        return obj

def listartists(ffilter=None, skip=None, limit=None):
    with session_scope() as session:
        query = session.query(Artist)
        if ffilter is not None:
            query = ffilter(query)
        query = limitoffset(query.order_by(Artist.name), skip, limit)
        qall = query.all()
        session.expunge_all()
        return qall

def listalbums(ffilter=None, skip=None, limit=None, order_by=(Artist.name, Album.year.desc())):
    with session_scope() as session:
        query = session.query(Album).join(Album.artist).options(contains_eager(Album.artist))
        if ffilter is not None:
            query = ffilter(query)
        if order_by is not None:
            query = limitoffset(query.order_by(*order_by), skip, limit)
        else:
            query = limitoffset(query, skip, limit)
        qall = query.all()
        session.expunge_all()
        return qall

def listalbumsbyartists(ffilter=None, skip=None, limit=None):
    with session_scope() as session:
        query = session.query(Artist).join(Artist.albums).options(contains_eager(Artist.albums, Album.artist))
        if ffilter is not None:
            query = ffilter(query)
        query = limitoffset(query.order_by(Artist.name, Album.year.desc()), skip, limit)
        qall = query.all()
        session.expunge_all()
        return qall

def listtracks(ffilter=None, skip=None, limit=None):
    with session_scope() as session:
        query = session.query(Track).join(Track.album).join(Album.artist).options(contains_eager(Track.album, Album.artist))
        if ffilter is not None:
            query = ffilter(query)
        query = limitoffset(query.order_by(Track.trackno), skip, limit)
        qall = query.all()
        session.expunge_all()
        return qall

def counttracks(ffilter=None):
    with session_scope() as session:
        query = session.query(Track)
        if ffilter is not None:
            query = ffilter(query)
        return query.count()
        
def countalbums(ffilter=None):
    with session_scope() as session:
        query = session.query(Album)
        if ffilter is not None:
            query = ffilter(query)
        return query.count()
        

def listtracksbyalbums(ffilter=None, skip=None, limit=None):
    with session_scope() as session:
        query = session.query(Album).join(Album.tracks).outerjoin(Track.genre).options(contains_eager(Album.tracks, Track.album, Album.artist), joinedload(Album.tracks, Track.genre))
        if ffilter is not None:
            query = ffilter(query)
        query = limitoffset(query.order_by(Album.year.desc(), Track.trackno), skip, limit)
        qall = query.all()
        # Force artist_name population
        for x in qall:
            for y in x.tracks:
                _ = y.album.artist.name
        session.expunge_all()
        return qall

def listtracksbyalbumsbyartists(ffilter=None, skip=None, limit=None):
    with session_scope() as session:
        query = session.query(Artist).join(Artist.albums).join(Album.tracks).options(contains_eager(Artist.albums, Album.tracks, Track.album))
        if ffilter is not None:
            query = ffilter(query)
        query = limitoffset(query.order_by(Artist.name, Album.year.desc(), Track.trackno), skip, limit)
        qall = query.all()
        # Force artist_name population
        for x in qall:
            for y in x.albums:
                for z in y.tracks:
                    _ = z.album.artist.name
        session.expunge_all()
        return qall

def searchartists(term, skip=None, limit=None):
    ffilter = lambda query: query.filter(Artist.name.contains(term))
    return listartists(ffilter, skip, limit)

def searchalbums(term, skip=None, limit=None):
    ffilter = lambda query: query.filter(Album.name.contains(term))
    return listalbums(ffilter, skip, limit)

def search(term, artists=True, albums=True, tracks=True, skip=None, limit=None):
    def ffilter(query):
        filters = []
        if artists:
            filters.append(Artist.name.contains(term))
        if albums:
            filters.append(Album.name.contains(term))
        if tracks:
            filters.append(Track.name.contains(term))
        if len(filters) > 0:
            query = query.filter(or_(*filters))
        return query
    return listtracksbyalbumsbyartists(ffilter, skip, limit)

if __name__ == "__main__":
    """
    print(listartists())
    print(listartists(skip=10, limit=20))
    print(listartists(lambda query: query.filter(Artist.name.contains('shall'))))
    
    print(listalbums())
    print(listalbums(skip=10, limit=20))
    print(listalbums(lambda query: query.filter(Album.name.contains('break'))))
    
    artists = listalbumsbyartists()
    print(artists)
    for artist in artists:
        print(artist.albums)
    
    artists = search('beyond', artists=True, albums=True, tracks=True)
    for artist in artists:
        print(artist)
        if 'albums' in artist.__dict__:
            for album in artist.albums:
                print(album)
                if 'tracks' in album.__dict__:
                    print(album.tracks)
    """
    
    print([x._asdict() for x in listtracks(skip=10, limit=20)])
    